src/vcpkg/binarycaching.cpp

```diff
@@ -1095,222 +1066,154 @@ namespace
+                          std::vector<std::string>&& write_prefixes)
+            : ObjectStorageProvider(paths, std::move(read_prefixes), std::move(write_prefixes))
         {
-            cmd.string_arg("--no-sign-request");
         }
-        return cmd_execute(cmd) == 0;
-    }
 
-    bool awscli_upload_file(const VcpkgPaths& paths,
-                            const std::string& aws_object,
-                            const Path& archive,
-                            const bool no_sign_request)
-    {
-        auto cmd =
-            Command{paths.get_tool_exe(Tools::AWSCLI)}.string_arg("s3").string_arg("cp").string_arg(archive).string_arg(
-                aws_object);
-        if (no_sign_request)
-        {
-            cmd.string_arg("--no-sign-request");
-        }
-        const auto out = cmd_execute_and_capture_output(cmd);
-        if (out.exit_code == 0)
+    private:
+        const std::string& vendor() const override { return m_vendor; }
+
+        bool stat(const std::string& url) const override
         {
-            return true;
+            auto cmd = Command{paths.get_tool_exe(m_tool)}.string_arg("-q").string_arg("stat").string_arg(url);
+            return cmd_execute(cmd) == 0;
         }
 
-        msg::println(Color::warning, msgAwsFailedToDownload, msg::exit_code = out.exit_code);
-        msg::write_unlocalized_text_to_stdout(Color::warning, out.output);
-        return false;
-    }
-
-    bool awscli_download_file(const VcpkgPaths& paths,
-                              const std::string& aws_object,
-                              const Path& archive,
-                              const bool no_sign_request)
-    {
-        auto cmd = Command{paths.get_tool_exe(Tools::AWSCLI)}
-                       .string_arg("s3")
-                       .string_arg("cp")
-                       .string_arg(aws_object)
-                       .string_arg(archive);
-        if (no_sign_request)
+        bool upload_file(const std::string& object, const Path& archive) const override
         {
-            cmd.string_arg("--no-sign-request");
+            auto cmd =
+                Command{paths.get_tool_exe(m_tool)}.string_arg("-q").string_arg("cp").string_arg(archive).string_arg(
+                    object);
+            const auto out = cmd_execute_and_capture_output(cmd);
+            if (out.exit_code == 0)
+            {
+                return true;
+            }
+
+            msg::println(Color::warning,
+                         msgObjectStorageToolFailedToDownload,
+                         msg::exit_code = out.exit_code,
+                         msg::tool_name = m_tool);
+            msg::write_unlocalized_text_to_stdout(Color::warning, out.output);
+            return false;
         }
-        const auto out = cmd_execute_and_capture_output(cmd);
-        if (out.exit_code == 0)
+
+        bool download_file(const std::string& object, const Path& archive) const override
```

Comment:
Feels like `upload_file()` and `download_file()` could be abstracted into `bool copy_file(source, destination)`.  But I'm not against this bit of repetition so not asking for changes at this time.